transaction/core/src/amount/v2.rs

```diff
@@ -0,0 +1,341 @@
+    /// TxOut shared secret.
+    ///
+    /// # Arguments
+    /// * `amount_shared_secret` - The shared secret, derived by hashing TxOut
+    ///   shared secret
+    pub fn get_value_from_amount_shared_secret(
+        &self,
+        amount_shared_secret: &[u8; 32],
+    ) -> Result<(Amount, Scalar), AmountError> {
+        let (expected_commitment, amount, blinding) = Self::compute_commitment(
+            self.masked_value,
+            &self.masked_token_id,
+            amount_shared_secret,
+        )?;
+        if self.commitment != expected_commitment {
+            // The commitment does not agree with the provided value and blinding.
+            // This either means that the commitment does not correspond to the shared
+            // secret, or that the amount is malformed (and is probably not
+            // spendable).
+            return Err(AmountError::InconsistentCommitment);
+        }
+
+        Ok((amount, blinding))
+    }
+
+    /// Compute the crc32 of the compressed commitment
+    pub fn commitment_crc32(&self) -> u32 {
+        Self::compute_commitment_crc32(&self.commitment)
+    }
+
+    /// Recovers an Amount from only the masked value and masked_token_id, and
+    /// shared secret.
+    ///
+    /// Note: This fails and produces gibberish if the shared secret is wrong.
+    ///
+    /// * You should confirm by checking against the real commitment, or the the
+    ///   crc32 of commitment.
+    ///
+    /// Arguments:
+    /// * masked_value: u64
+    /// * masked_token_id: &[u8], either 0 or 4 bytes
+    /// * shared_secret: The shared secret curve point
+    ///
+    /// Returns:
+    /// * MaskedAmount
+    /// * Amount (token id and value)
+    /// or
+    /// * An amount error
+    pub fn reconstruct(
+        masked_value: u64,
+        masked_token_id: &[u8],
+        tx_out_shared_secret: &RistrettoPublic,
+    ) -> Result<(Self, Amount), AmountError> {
+        let amount_shared_secret = get_amount_shared_secret(tx_out_shared_secret);
+
+        let (expected_commitment, amount, _) =
+            Self::compute_commitment(masked_value, masked_token_id, &amount_shared_secret)?;
+
+        let result = Self {
+            commitment: expected_commitment,
+            masked_value,
+            masked_token_id: masked_token_id.to_vec(),
+        };
+
+        Ok((result, amount))
+    }
+
+    /// Compute the expected commitment corresponding to a masked value, masked
+    /// token id, and shared secret, returning errors if the masked token id
+    /// is malformed.
+    fn compute_commitment(
+        masked_value: u64,
+        masked_token_id: &[u8],
+        amount_shared_secret: &[u8; 32],
+    ) -> Result<(CompressedCommitment, Amount, Scalar), AmountError> {
+        let (value_mask, token_id_mask, blinding) = get_blinding_factors(amount_shared_secret);
+
+        let token_id = TokenId::from(match masked_token_id.len() {
+            0 => 0,
```

Comment:
~:question: My understanding is that v2 of the `VersionedMasked` is block 3, and token id was added in block 2, do we anticipate token id being 0 with this change?~

:x: Thinking about this more in `new()` the `amount.token_id` is *always* masked with `token_id_mask`.  I think of `0` got in here somehow this would not provide the right `token_id`

Reply:
I guess you're right, the 0 byte payload is the backwards compat path for before token id existed. But that compat path is only needed when the masked amount is version 1. A version 2 masked amount will never have an empty token id field. So we could safely remove this, it's leftover from copy paste. Thanks